### PR TITLE
HBASE-23035 Retain region to the last RegionServer make the failover …

### PR DESCRIPTION
**Pull request by infraio**
…slower (addendum)

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 33s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 4 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 32s |  master passed  |
| :green_heart: |  compile  |   0m 54s |  master passed  |
| :green_heart: |  checkstyle  |   1m 18s |  master passed  |
| :green_heart: |  shadedjars  |   4m 31s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 36s |  master passed  |
| :blue_heart: |  spotbugs  |   4m  6s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m  4s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m  1s |  the patch passed  |
| :green_heart: |  compile  |   0m 56s |  the patch passed  |
| :green_heart: |  javac  |   0m 56s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 31s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m  8s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  18m  8s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 12s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 162m  5s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   | 221m 52s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.procedure.TestSCPWithoutMetaWithReplicas |
|   | hadoop.hbase.master.procedure.TestSCPWithMetaWithReplicas |
|   | hadoop.hbase.master.procedure.TestSCPWithMetaWithReplicasWithoutZKCoordinated |
|   | hadoop.hbase.master.procedure.TestSCPWithoutMetaWithReplicasWithoutZKCoordinated |
|   | hadoop.hbase.master.procedure.TestSCPWithReplicasWithoutZKCoordinated |
|   | hadoop.hbase.tool.TestBulkLoadHFilesSplitRecovery |
|   | hadoop.hbase.tool.TestSecureBulkLoadHFilesSplitRecovery |
|   | hadoop.hbase.master.procedure.TestSCPWithReplicas |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/652 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 60ebcc974ba1 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-652/out/precommit/personality/provided.sh |
| git revision | master / 537c5f0660 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/1/testReport/ |
| Max. process+thread count | 4427 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   2m 11s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 6 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   8m 17s |  master passed  |
| :green_heart: |  compile  |   1m 17s |  master passed  |
| :green_heart: |  checkstyle  |   1m 44s |  master passed  |
| :green_heart: |  shadedjars  |   6m  4s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 49s |  master passed  |
| :blue_heart: |  spotbugs  |   5m 23s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   5m 20s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   6m 54s |  the patch passed  |
| :green_heart: |  compile  |   1m 17s |  the patch passed  |
| :green_heart: |  javac  |   1m 17s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 33s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   6m  8s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  23m 48s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 43s |  the patch passed  |
| :green_heart: |  findbugs  |   5m 15s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 233m 35s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 25s |  The patch does not generate ASF License warnings.  |
|  |   | 314m  9s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/652 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 1bbacb5d7816 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-652/out/precommit/personality/provided.sh |
| git revision | master / d3c1bb42e9 |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/2/testReport/ |
| Max. process+thread count | 4805 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java [hunk 1]
```diff
@@ -172,11 +172,11 @@ protected boolean waitInitialized(MasterProcedureEnv env) {
 
   private void queueAssign(MasterProcedureEnv env, RegionStateNode regionNode)
       throws ProcedureSuspendedException {
-    // Here the assumption is that, the region must be in CLOSED state, so the region location
-    // will be null. And if we fail to open the region and retry here, the forceNewPlan will be
-    // true, and also we will set the region location to null.
     boolean retain = false;
-    if (!forceNewPlan) {
+    if (forceNewPlan) {
```
Comment (ramkrish86) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java]: Have you done any other commit prior to this change? Seems currently TransitRegionStateProcedure does not get called with forceNewPlan = true.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java [hunk 2]
```diff
@@ -399,6 +399,7 @@ public void reportTransition(MasterProcedureEnv env, RegionStateNode regionNode,
   public void serverCrashed(MasterProcedureEnv env, RegionStateNode regionNode,
       ServerName serverName) throws IOException {
     // force to assign to a new candidate server
+    // AssignmentManager#regionClosedAbnormally will set region location to null
     // TODO: the forceNewPlan flag not be persistent so if master crash then the flag will be lost.
     // But assign to old server is not big deal because it not effect correctness.
     // See HBASE-23035 for more details.
```
Comment (ramkrish86) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java]: Already some commit has been done for this JIRA?

Comment (ramkrish86) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java]: I updated and saw some thing has already been committed. Will check once again.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/procedure/TestSCPBase.java [hunk 3]
```diff
@@ -141,36 +135,6 @@ protected long getSCPProcId(ProcedureExecutor<?> procExec) {
     return procExec.getActiveProcIds().stream().mapToLong(Long::longValue).min().getAsLong();
   }
 
-  private void assertReplicaDistributed(Table t) throws IOException {
```
Comment (ramkrish86) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/master/procedure/TestSCPBase.java]: This you will come back later after the other new JIRA for load balancer is fixed?

**Comment by ramkrish86** [comment]
+1.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/procedure/TestSCPBase.java (hunk 3 above)
Comment (infraio) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/procedure/TestSCPBase.java]: Yes, add it back in HBASE-23078.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java (hunk 1 above)
Comment (infraio) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/TransitRegionStateProcedure.java]: Please pull the latest commit. SCP will create TRSP with forceNewPlan = true.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   3m 30s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 7 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 38s |  master passed  |
| :green_heart: |  compile  |   0m 54s |  master passed  |
| :green_heart: |  checkstyle  |   1m 17s |  master passed  |
| :green_heart: |  shadedjars  |   4m 46s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 40s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 11s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m  9s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 11s |  the patch passed  |
| :green_heart: |  compile  |   0m 59s |  the patch passed  |
| :green_heart: |  javac  |   0m 59s |  the patch passed  |
| :broken_heart: |  checkstyle  |   1m 18s |  hbase-server: The patch generated 3 new + 2 unchanged - 0 fixed = 5 total (was 2)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 48s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  16m 33s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 38s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 38s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 331m 56s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 36s |  The patch does not generate ASF License warnings.  |
|  |   | 394m 11s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |
|   | hadoop.hbase.tool.TestSecureBulkLoadHFiles |
|   | hadoop.hbase.replication.TestReplicationKillSlaveRSWithSeparateOldWALs |
|   | hadoop.hbase.client.TestSnapshotTemporaryDirectory |
|   | hadoop.hbase.replication.TestReplicationSmallTestsSync |
|   | hadoop.hbase.tool.TestBulkLoadHFiles |
|   | hadoop.hbase.client.TestAsyncTableAdminApi |
|   | hadoop.hbase.regionserver.TestRegionServerAbortTimeout |
|   | hadoop.hbase.replication.TestReplicationDisableInactivePeer |
|   | hadoop.hbase.replication.TestReplicationSmallTests |
|   | hadoop.hbase.client.TestCloneSnapshotFromClientNormal |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/652 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 1a2feb9e656a 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-652/out/precommit/personality/provided.sh |
| git revision | master / 3250a8021a |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/3/artifact/out/diff-checkstyle-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/3/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/3/testReport/ |
| Max. process+thread count | 4950 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   3m 37s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 7 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   6m  2s |  master passed  |
| :green_heart: |  compile  |   1m  0s |  master passed  |
| :green_heart: |  checkstyle  |   1m 29s |  master passed  |
| :green_heart: |  shadedjars  |   5m  3s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 40s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 42s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 40s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 44s |  the patch passed  |
| :green_heart: |  compile  |   1m  0s |  the patch passed  |
| :green_heart: |  javac  |   1m  0s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 28s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m  1s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  18m  1s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 36s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 35s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 230m 27s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 27s |  The patch does not generate ASF License warnings.  |
|  |   | 296m 43s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/652 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 9b9c2baf32a3 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-652/out/precommit/personality/provided.sh |
| git revision | master / 3250a8021a |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/4/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/4/testReport/ |
| Max. process+thread count | 4718 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-652/4/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**